src/graphql/data/TopTokens.ts

```diff
@@ -129,63 +130,108 @@ function toContractInput(token: PrefetchedTopToken) {
+
+  const environment = useRelayEnvironment()
 
   // TopTokens should ideally be fetched with usePaginationFragment. The backend does not current support graphql cursors;
   // in the meantime, fetchQuery is used, as other relay hooks do not allow the refreshing and lazy loading we need
-  const prefetchedSelectedTokens = useFilteredTokens(useSortedTokens(prefetchedData.topTokens))
-  const contracts: ContractInput[] = useMemo(
-    () => prefetchedSelectedTokens.slice(page * PAGE_SIZE, (page + 1) * PAGE_SIZE).map(toContractInput),
-    [page, prefetchedSelectedTokens]
+  const loadTokensWithPriceHistory = useCallback(
+    (contracts: ContractInput[], onSuccess: (data: TopTokens_TokensQuery['response'] | undefined) => void) => {
+      fetchQuery<TopTokens_TokensQuery>(
+        environment,
+        tokensQuery,
+        { contracts, duration },
+        { fetchPolicy: 'store-or-network' }
+      )
+        .toPromise()
+        .then(onSuccess)
+    },
+    [duration, environment]
   )
 
-  useEffect(() => {
-    const subscription = fetchQuery<TopTokens_TokensQuery>(
-      environment,
-      tokensQuery,
-      { contracts, duration },
-      { fetchPolicy: 'store-or-network' }
-    ).subscribe({
-      start() {
-        setLoading(true)
-      },
-      complete() {
+  const loadMoreTokens = useCallback(() => {
+    setLoading(true)
+    const contracts = prefetchedSelectedTokensWithoutPriceHistory
+      .slice(page * PAGE_SIZE, (page + 1) * PAGE_SIZE)
+      .map(toContractInput)
+    loadTokensWithPriceHistory(contracts, (data) => {
+      if (data?.tokens) {
+        data.tokens.forEach((token) =>
+          !!token ? (tokensWithPriceHistoryCache[`${token.chain.toString()}${token.address}`] = token) : null
+        )
+        setTokens([...(tokens ?? []), ...data.tokens])
         setLoading(false)
-      },
-      next(data) {
-        appendTokens(data.tokens as TopToken[])
-      },
+        setPage(page + 1)
+      }
     })
-    return subscription.unsubscribe
-  }, [appendTokens, contracts, duration, environment])
+  }, [prefetchedSelectedTokensWithoutPriceHistory, page, loadTokensWithPriceHistory, tokens])
 
-  return { loading, tokens: useFilteredTokens(useSortedTokens(tokens)) as TopToken[], loadMoreTokens }
+  // Reset count when filters are changed
+  useLayoutEffect(() => {
+    const { everyTokenHasBeenPreloaded, preloadedSelectedTokens } = checkIfAllTokensCached(
+      prefetchedSelectedTokensWithoutPriceHistory
+    )
+    if (everyTokenHasBeenPreloaded) {
+      setTokens([...preloadedSelectedTokens])
+      setLoading(false)
+      return
+    } else {
+      setLoading(true)
+      setTokens([])
+      const contracts = prefetchedSelectedTokensWithoutPriceHistory.slice(0, PAGE_SIZE).map(toContractInput)
+      loadTokensWithPriceHistory(contracts, (data) => {
+        if (data?.tokens) {
+          data.tokens.forEach((token) =>
```

Comment:
why do you mutate the array vs map here? would that fix the TS error?

Reply:
good idea, changed to map. that doesn't fix the ts error, but i spread the array instead and it fixed it